lib/systemd/user-level-services.nix

```diff
@@ -0,0 +1,105 @@
+{ nixpkgs
+, system ? "x86_64-linux"
+, pkgs ? nixpkgs.legacyPackages.${system}, ... }:
+let
+  mkVM = moduleConfig: serviceConfig: nixpkgs.lib.nixosSystem {
+    inherit system;
+    modules = [ ({ ... }: {
+
+      imports = [ moduleConfig.module ];
+
+      services.${moduleConfig.nixOsServiceName} = {
+        enable = true;
+      } // serviceConfig;
+
+    }) ];
+  };
+
+in {
+  mkActivateScript = activateScriptName:
+      { module, nixOsServiceName, primaryService, primaryPortOptionName, auxiliaryServices}@moduleConfig:
+      { backupAction ? "", checkAction ? null, restoreAction ? ""}:
+      serviceConfig:
+    let
+      vm = mkVM moduleConfig serviceConfig;
+      allServices = [primaryService] ++ auxiliaryServices;
+    in
+    pkgs.writeShellScriptBin activateScriptName ''
+      set -euo pipefail
+      export XDG_RUNTIME_DIR="/run/user/$UID"
+      mkdir -v -p "$HOME/.config/systemd/user/default.target.wants"
+      ${if backupAction != ""
+      then
+        ''
+          # Deploy may change DB schema via migration and we should be able
+          # to restore its state to a point before an update to successfully rollback
+          # failed deployment
+          echo Stopping ${primaryService}
+          systemctl --user stop ${primaryService} || true
```

Comment:
I still don't quite get why `systemctl stop` has to always be run prior to the backup in arbitrary case

Reply:
Until now the implementation was to stop the service before backup.

Reply:
removed the service stop